Scheduler flexibility with Multiplexer Load Balancer

vrann opened the pull request:
<!--- Provide a concise summary of your changes in the Title -->

## Description
This change allows to select custom Load Balancing strategies based on the annotation of the action. I introduced new MuxBalancer. When action is created, developer can specify "activationStrategy" annotation. If OpenWhisk is configured to use MuxBalancer, it will pick up annotation value and will look in the map of the configured activation strategies for that alias. In case if alias is found, it will delegate the publishing of the activation to the LoadBalancer corresponding to the alias. Otherwise, MuxBalancer will use the "default" Load Balancer to delegate the publication.

The configuration of the MuxBalancer looks like folloving:

```
whisk.loadbalancer {
    strategy {
      default = "fully.qualified.class.name.for. LoadBalancer0"
      custom = {
           "activationStrategyName1" = "fully.qualified.class.name.for. LoadBalancer1"
           "activationStrategyName2" = "fully.qualified.class.name.for. LoadBalancer2"
            ...
      }
    }
}
```

developer would use it in the following way:
`wsk action create hello hello.js --annotation activationStrategy activationStrategyName1`

### Why this change is important for us

When working on the new mechanism of processing of the activations we haven't found better extension point on where it could be plugged in to the system other than changing the logic of Load Balancer. This motivates us to introduce a mechanism which would allow to select the Load Balancers based on some characteristics of activations. We considered kind as one possibility for such characteristic, and it could be added later. For now, annotation looks like the most straightforward way (with less drawbacks)

### Breaking change?
It is not a breaking change, because the support for activation strategy annotation would be enabled only if  MuxBalancer is selected as a default Load Balancer. And only in that case the `strategy` node of the configuration need to be configured. Otherwise system does not use it.

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [ ] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

vrann committed:
OW-5097: Implement LoadBalancer Strategy based on annotation

- implemented MuxLoadBalancer to dynamically select load balancing strategy based on annotation of the action

KeonHee committed:
[New Scheduler] Etcd installation & Implements EtcdClient (#5031)

* Implements EtcdClient

* Add license header

* Add configuration for tests

* Exclude etcd config from controller & invoker

* Add config for etcd test

* Separate etcd test from whisk unit test

* Apply scala compilation

dgrove-oss committed:
remove deprecated python:2 kind (#5064)

Python 2 EOL was more than a year ago; finish the cleanup and
remove the python:2 kind from the core repository.

dgrove-oss committed:
update unicode test dat files for runtime language levels (#5066)

KeonHee committed:
[New Scheduler] CI for testing related new scheduler (#5068)

* Add task for testing related to the scheduler

* Add Scheduler to request template

* Add travis job for testing scheduler

jiangpengcheng committed:
[New Scheduler] Add a centralized watcher for etcd data (#5069)

* Add a centralized watcher for etcd data

* Fix class name format

* Fix compilation error

* Enable test for WatcherService

akrabat committed:
Add php:8.0 kind (#5071)

Also, set example PHP runtime to v7.4.

fe-lix- committed:
Make cache expiration time configurable (#5076)

jbampton committed:
chore: fix spelling (#5077)

KeonHee committed:
[New Scheduler] Implement KeepAliveService (#5067)

* Add KeepAliveService

* Add KeepAliveService to Scheduler

* Update name for case class

* Include workers in state data to thread safe

jiangpengcheng committed:
[New Scheduler] Add container counter (#5072)

* [New Scheduler] Add container counter

Get container count from ETCD when related data get updated in ETCD

* Fix tests

* Fix tests

jbampton committed:
fix: Python 3 urlencode (#5082)

The urllib.urlencode() function is now urllib.parse.urlencode()

joni-jones commented:
One more thing that should be changed from OW in standalone mode. The `KafkaLauncher` uses `LeanBalancer` which should be replaced by `MuxBalancer`. Something like this https://github.com/adobe/openwhisk/pull/4/files?file-filters%5B%5D=.conf&file-filters%5B%5D=.gradle&file-filters%5B%5D=.scala#diff-f599539516ceae141ad2c644fcf43b5edc3186e21ff6a828dc67d5adabd0109aR115-R119.

YevSent reviewed (changes requested):
One more thing that should be changed from OW in standalone mode. The KafkaLauncher uses LeanBalancer which should be replaced by MuxBalancer. Something like this https://github.com/adobe/openwhisk/pull/4/files?file-filters%5B%5D=.conf&file-filters%5B%5D=.gradle&file-filters%5B%5D=.scala#diff-f599539516ceae141ad2c644fcf43b5edc3186e21ff6a828dc67d5adabd0109aR115-R119.

dgrove-oss committed:
configure more github properties via asf.yaml (#5086)

rabbah committed:
Change up travis URL. (#5087)

jbampton committed:
fix: update Homebrew commands (#5081)

- brew cask install -> brew install --cask
- brew tap caskroom/cask -> brew tap homebrew/cask

style95 committed:
[New Scheduler] Add DataManagementService (#5063)

* Add DataManagementService

* Update common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala

Apply comment

Co-authored-by: Brendan Doyle <bdoyle0182@gmail.com>

* Apply suggestions from code review

Update comments

Co-authored-by: Brendan Doyle <bdoyle0182@gmail.com>

* Apply comments

* Add unit tests for DataManagementService

* Remove unused variable

* Add the license header

* Change Lease

* Pull docker image for the api gateway in advance

Co-authored-by: Brendan Doyle <bdoyle0182@gmail.com>

ningyougang committed:
Implement InvokerHealthyManager (#5061)

ishaanthakur committed:
Fixed link for Ansible (#5099)

rabbah committed:
Quote ansible default value. (#5101)

upgle committed:
[New Scheduler] Add ActivationService (#5070)

* Add ActivationService for scheduler

* Add annotation

* Add license header

* Reformat activation.proto

* Reduce request timeout

* Add license header

* Scan code before compiling the code

Commit:
Updated Kamon bundle dependencies to 2.1.12 version (#5085)

dgrove-oss committed:
change default nodejs kind from nodejs:10 to nodejs:14

vrann committed:
DEVX-533: Implement LoadBalancer Strategy based on annotation

- switched on MuxBalancer for the standalone

dgrove-oss committed:
update default python3 image name (#5104)

We stopped building the old (non-actionloop) python 3 docker image
about six months ago. Update the default image name so we use/test the
actionloop-based python3 runtime.

dgrove-oss committed:
attempt to hack around docker rate limiting during tests

ningyougang committed:
Make kafka version configurable (#5046)

ningyougang committed:
[New Scheduler]Implement PFCInvokerServer (#5098)

* Implement PFCInvokerServer

* Remove /memory api

* Default implement enable/disable

ningyougang committed:
Add prefix for topics (#5062)

- Add prefix for topics
- Add extra prefix for userEvent topic only

dgrove-oss committed:
deprecate nodejs:10 kind

ningyougang committed:
[New Scheduler] Implement FunctionPullingContainerPool (#5102)

* Implement FunctionPullingContainerPool

* Fix review points

dgrove-oss committed:
prefetch openwhisk/example to avoid timeout during test execution

dgrove-oss committed:
increase default totalWait when getting an activation

dgrove-oss committed:
refresh module list (#5117)

KeonHee committed:
[New Scheduler] Manage container creation (#5074)

* Manage container creation

* Add ContainerManager Test

* Add dedicated namespace

* Remove namespace

* Apply scala fmt

* Add dedicatedNamespaces filter

* Add dedicatedNamespaces test

* Move InvokerState to common

* Unify InvokerHealth message

* Add configuration for test

* Add license header

* Remove compare InvokerResourceMessage

upgle committed:
[New Scheduler] Add container message consumer (#5111)

* Add container message consumer

* Reformat code

* Fix test case error

Co-authored-by: ning.yougang <ning.yougang@navercorp.com>

ningyougang committed:
Document for prewarmed container (#4910)

vrann committed:
#5120: Unit Tests failing due to testcontainers (#5121)

- upgraded the version of testcontainers

ningyougang committed:
[New Scheduler] Implement FunctionPullingContainerProxy (#5113)

* Implement FunctionPullingContainerProxy

* Fix type mismatch error

* Fix cannot find Jsonwriter or JsonFormat type

* Fix Unit companion object is not allowed in source

* Add missing resourceTags and imagePullSecret

* Make test case more stable

* Remove unnecessary feature

style95 committed:
Update the notice year (#5122)

vrann committed:
#5060: Upgrade to Akka 2.6.12 (#5065)

- removed usage implicit custom ActorMaterializer
- upgraded to Akka Http 10.2.3
- changed the way how Https set up when host name verification is disabled
- changed RestartSink.withBackoff to accept RestartSettings
- changed akka.actor.FSM.setTimer => akka.actor.FSM.startTimerAtFixedRate
- changed akka.actor.Scheduler.schedule => akka.actor.Scheduler.scheduleAtFixedRate
- updated to the new signature of Source.actorRef
- replaced deprecated HttpResponse.copy
- replaced Gzip with Coders.Gzip
- fixed the scalafmt in tests
- removed the implicit Materializer shutdown considering its lifecycle is managed by akka
- increased the ansible wait time
- switched to classic networking
- disabled JFR
- merging with New Scheduler
- increased wait time for FSM due to intermittent failures in build

jiangpengcheng committed:
[New Scheduler]Add CreationJobManager (#5116)

* Add CreationJobManager

* Remove unused import

* Use finite duration string for a config

* Fix tests

* Resolve rebase conflicts

* Fix tests

ningyougang committed:
Implement ActivationClientProxy (#5119)

ActivationClientProxy actor fetches activationMessage from scheudler side
using akka grpc, and forwards the activationMessage to its parent actor
FunctionPullContainerProxy actor.

ningyougang committed:
Fix build error (#5123)

jiangpengcheng committed:
Implement MongoDBArtifactStore (#4963)

* Implement MongoDBActivationStore

Co-authored-by: Chetan Mehrotra <chetanm@apache.org>

* Upgrade mongo-scala to 2.7.0

* Fix test

* Add license headers

* Add default value for mongodb_connect_string

* Rename graph stage classes used for mongodb gridfs

* Update readme

* Update based on comments

* Fix typo and update README

* Rename db.backend to db.artifact_store.backend

Co-authored-by: Chetan Mehrotra <chetanm@apache.org>

jiangpengcheng committed:
Fix deprecated error (#5128)

* Fix deprecated error

* Make testcase stable

dgrove-oss committed:
update modules.md to move deploy-mesos and external-resources to inactive (#5127)

dgrove-oss committed:
update to openwhisk-client-js 3.21.4 (#5129)

ddragosd committed:
Docs update for local dev (#5134)

ryutoyasugi committed:
Fix ansible README (#5135)

rabbah committed:
Remove Mesos container factory. (#5133)

Remove Mesos container factory and related library, tests, and documentation

KeonHee committed:
[New Scheduler] Manage memory queues in scheduler (#5118)

* Add QueueManager

* Add test configuration

* Remove deprecated functions

* Fix memory queue test

* Change string to FiniteDuration

vrann committed:
Merge branch 'master' into DEVX-533-2
